tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs

```diff
@@ -0,0 +1,122 @@
+namespace VstsPipelineSync
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Data.SqlClient;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using DevOpsLib;
+    using DevOpsLib.VstsModels;
+
+    public class VstsBuildBatchUpdate
+    {
+        readonly DevOpsAccessSetting devOpsAccessSetting;
+        readonly string dbConnectionString;
+
+        public VstsBuildBatchUpdate(DevOpsAccessSetting devOpsAccessSetting, string dbConnectionString)
+        {
+            ValidationUtil.ThrowIfNull(devOpsAccessSetting, nameof(devOpsAccessSetting));
+
+            this.devOpsAccessSetting = devOpsAccessSetting;
+            this.dbConnectionString = dbConnectionString;
+        }
+
+        public async Task RunAsync(TimeSpan waitPeriodAfterEachUpdate, CancellationToken ct)
+        {
+            var buildManagement = new BuildManagement(devOpsAccessSetting);
+            var lastUpdatePerBranchPerDefinition = new Dictionary<string, Dictionary<BuildDefinitionId, DateTime>>();
+
+            while (!ct.IsCancellationRequested)
+            {
+                try
+                {
+                    Console.WriteLine($"Import Vsts Builds data started at {DateTime.UtcNow}");
+
+                    lastUpdatePerBranchPerDefinition.Upsert("refs/heads/master", await ImportVstsBuildsDataAsync(buildManagement, "refs/heads/master", BuildExtension.MasterBranchBuildDefinitions,  lastUpdatePerBranchPerDefinition));
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"Unexcepted Exception: {ex}");
+                }
+
+                Console.WriteLine($"Import Vsts Builds data finished at {DateTime.UtcNow}; wait {waitPeriodAfterEachUpdate} for next update.");
+                await Task.Delay((int)waitPeriodAfterEachUpdate.TotalMilliseconds);
+            }
+        }
+
+        private async Task<Dictionary<BuildDefinitionId, DateTime>> ImportVstsBuildsDataAsync(BuildManagement buildManagement, string branch, HashSet<BuildDefinitionId> buildDefinitionIds, Dictionary<string, Dictionary<BuildDefinitionId, DateTime>> lastUpdatePerBranchPerDefinition)
+        {
+            Dictionary<BuildDefinitionId, DateTime> lastUpdatePerDefinition = lastUpdatePerBranchPerDefinition.GetIfExists(branch);
+            if (lastUpdatePerDefinition == null)
+            {
+                lastUpdatePerDefinition = new Dictionary<BuildDefinitionId, DateTime>();
+            }
+
+            foreach (BuildDefinitionId buildDefinitionId in buildDefinitionIds)
+            {
+                DateTime maxLastChange = await ImportVstsBuildsDataForSpecificDefinitionAsync(buildManagement, branch, buildDefinitionId, lastUpdatePerDefinition.GetIfExists(buildDefinitionId));
```

Comment:
I'm assuming this will default to DateTime.MinValue according to link?
https://stackoverflow.com/questions/13957701/new-datetime-vs-defaultdatetime

Reply:
yes, correct.